Storage Wars

cicdw opened the pull request:
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
This PR builds off @joshmeek 's work and beefs up our `Storage` classes as follows:
- adds PIN 7 (closes #921 )
- adds a new `Memory` storage option for local testing
- implements a new and expanded `Storage` interface for storing multiple flows, and accessing flows contained in Storage
- adds `storage` submodule to our exposed API documentation

I also removed a few things from the initial implementation for the sake of simplicity / exposing a cleaner interface until we know how we want these things to be used.

Assuming this PR is approved and merged, it would be very easy from here to implement a `LocalEnvironment` and a `DaskClusterEnvironment` using this interface.  I avoided taking it that far because it's already a large PR.

For the full details of what this PR is trying to achieve, please read the PIN-7 that is included.

joshmeek commented on src/prefect/environments/storage/docker.py:
```diff
@@ -279,78 +355,14 @@ def create_dockerfile_object_from_base_image(
 
             dockerfile.write(file_contents)
 
-    def create_dockerfile_object_from_dockerfile(
```
What's the intention behind removing the entire full dockerfile option? Too many extra required steps i.e. healthcheck, copying flows, etc...?

cicdw replied:
So the intent was to simplify the interface for now, and the _only_ difference between `create_dockerfile_object_from_dockerfile` and the current method is that the current method's docker file is:
```
FROM {base_image}

# DO pip stuff
```
so I figured when we need this feature we can easily generalize our current method (and avoid having two methods).

joshmeek commented on src/prefect/environments/storage/memory.py:
```diff
@@ -0,0 +1,96 @@
+            - Union[Flow, FlowRunner]: the requested Flow or a FlowRunner for the requested Flow
+        """
+        if not flow_location in self.flows:
+            raise ValueError("Flow is not contained in this Storage")
```
Should this `ValueError` go in the docstring?

joshmeek commented on src/prefect/environments/storage/memory.py:
```diff
@@ -0,0 +1,96 @@
+            - str: the location of the newly added flow in this Storage object
+        """
+        if flow in self:
+            raise ValueError(
```
Should this `ValueError` go in the docstring?

joshmeek commented on src/prefect/environments/storage/memory.py:
```diff
@@ -0,0 +1,96 @@
+    from prefect.engine.flow_runner import FlowRunner
+
+
+class Memory(Storage):
```
At first I didn't understand how this worked but upon seeing the `build` function it became clear. Nice! May want to update the class's main docstring because it wasn't that obvious as to what it meant by `Memory`. (also just could be because I haven't had a coffee yet this morning)

joshmeek commented on src/prefect/environments/storage/bytes.py:
```diff
@@ -1,24 +0,0 @@
-import prefect
```
Ah makes sense, memory kind of replaces it for now. Do you think we should still plan on making a bytes related storage?

cicdw replied:
Yea absolutely, I imagined its replacement would be `LocalStorage` which stores `bytes` to local filesystem.

joshmeek commented:
This is awesome!

joshmeek commented on src/prefect/serialization/storage.py:
```diff
@@ -18,12 +18,11 @@ class Meta:
     registry_url = fields.String(allow_none=True)
     image_name = fields.String(allow_none=True)
     image_tag = fields.String(allow_none=True)
-    flow_file_path = fields.String(allow_none=True)
```
So are these going to be inferred now based on name?

cicdw replied:
Nope, oops; I need to add this attribute to flow serialization.

cicdw replied:
Addressed by serializing `flows` attribute on storage. 👍

joshmeek commented on src/prefect/environments/storage/base.py:
```diff
@@ -1,34 +1,99 @@
         pass
 
-    def build(self, flow: "prefect.Flow") -> "Storage":
+    def get_env_runner(self, flow_location: str) -> Any:
```
Could you also add a note that this is a way to control a flow run based on environment variables. Since it just says "for running the flow" maybe it should have something in tune with 

`which accepts a collection of environment variables for running the flow. (e.g. specifying an executor with PREFECT__ENGINE__EXECUTOR__etc...)`

jlowin replied:
*that* accepts ;)

jlowin commented on CHANGELOG.md:
```diff
@@ -6,13 +6,14 @@ These changes are available in the [master branch](https://github.com/PrefectHQ/
 ### Features
 
-- Add new `Storage` and `Environment` specifications - [#936](https://github.com/PrefectHQ/prefect/pull/936)
+- Add new `Storage` and `Environment` specifications - [#936](https://github.com/PrefectHQ/prefect/pull/936), [#956](https://github.com/PrefectHQ/prefect/pull/936)
```
```suggestion
- Add new `Storage` and `Environment` specifications - [#936](https://github.com/PrefectHQ/prefect/pull/936), [#956](https://github.com/PrefectHQ/prefect/pull/956)
```

cicdw commented on docs/guide/PINs/PIN-7-Storage-Execution.md:
```diff
@@ -0,0 +1,48 @@
+- a `build() -> Storage` method for "building" the storage
+- a `serialize() -> dict` method for serializing the relevant information about this Storage for later re-use.
+
+Note that there is some weird behavior that could be possible if a user adds a Flow to Storage, serializes the Storage, and then deserializes in a new process.  For this reason, we might also consider a `load()` method which will allow the Storage unit to inspect itself and populate `self.flows`.
```
Yea, so take `Docker` storage for instance: suppose I add two flows to the storage, and run `.build()` then `.serialize()`.  Suppose I then _deserialize_ in a new process.  I then attempt to run `get_flow_location` for a flow.  An error will be raised that this flow is not located in this storage, even though it actually is.

cicdw replied:
I have a way to fix this.  Will be committing shortly.

jlowin replied:
This is what I had in mind about string keys rather than flow objects, I think

cicdw replied:
Yup, I've updated the PR to now serialize `self.flows` for Docker storage as a dictionary of name -> location, that way the flow doesn't actually have to serialize it's own location.

cicdw commented on src/prefect/environments/storage/docker.py:
```diff
@@ -20,53 +21,41 @@ class Docker(Storage):
         image_tag: str = None,
         env_vars: dict = None,
         files: dict = None,
-        flow_file_path: str = "/root/.prefect/flow_env.prefect",
```
Yea, for now, although that should be easy to change if we want.

cicdw commented:
@jlowin happy to update the signature of that method.  Just for the record, the only reason I left it like that was in case new types of storage relied on other permanent attributes of Flows other than name, but I do think it's better to keep it simple for now.

jlowin commented:
This is such a clean implementation and advances so many of our goals that we thought would take longer, so let's start with the simpler version to get off the ground

cicdw commented:
@jlowin and @joshmeek : I've addressed all of your feedback and updated `Docker` storage to serialize its `flows` attribute.  That way, knowledge of containment survives serialization of flows / storage objects without any additional work!  Additionally, all the agent will have to do when it picks up a flow run is something like:
```python
environment = flow.environment
storage = flow.storage
environment.setup(storage)
environment.execute(storage, storage.flows[flow.name])
```

cicdw commented:
Actually, writing taht makes me wonder whether `get_flow_location` is even a necessary method at this point, since the `flows` attribute has been elevated.  Thoughts?
